### PR TITLE
feat(mqtt5): resume session on reconnect

### PR DESCRIPTION
**Pull request by jcosentino11**
**Issue #, if available:**

**Description of changes:**
When MQTT5 client purposefully reconnects, do so with cleanSession=false

**Why is this change necessary:**
To match behavior of MQTT3 client, and improve performance when reconnecting (subscriptions don't need to be made again).

**How was this change tested:**
- [x] Updated or added new unit tests.
- [ ] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [ ] I confirm that the change is backwards compatible.
- [ ] Any modification or deletion of public interfaces does not impact other plugin components.
- [ ] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/mqttclient/AwsIotMqtt5Client.java [hunk 1]
```diff
@@ -380,7 +389,7 @@ public void close() {
     public void reconnect(long timeoutMs) throws TimeoutException, ExecutionException, InterruptedException {
         logger.atInfo().log("Reconnecting MQTT client most likely due to device configuration change");
         disconnect().get(timeoutMs, TimeUnit.MILLISECONDS);
-        connect().get(timeoutMs, TimeUnit.MILLISECONDS);
+        connect(true).get(timeoutMs, TimeUnit.MILLISECONDS);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqttclient/AwsIotMqtt5Client.java:392]: You don't actually know that we ever successfully connected at this point. Store an atomic boolean to represent if we've ever successfully connected which is set in the successful connection callback.
Then read the flag when connecting.

Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/mqttclient/AwsIotMqtt5Client.java:392]: ah right, looks like theres one that exists already, will rework

File: src/main/java/com/aws/greengrass/mqttclient/AwsIotMqtt5Client.java [hunk 2]
```diff
@@ -71,6 +71,7 @@ class AwsIotMqtt5Client implements IndividualMqttClient {
     private static final String RESUB_LOG_EVENT = "resubscribe";
     private final Provider<AwsIotMqtt5ClientBuilder> builderProvider;
 
+    @Getter // for testing
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/mqttclient/AwsIotMqtt5Client.java:74]: Keep it at package visibility if able.

**Review by MikeDombo — APPROVED** [review]

File: src/main/java/com/aws/greengrass/mqttclient/AwsIotMqtt5Client.java (hunk 2 above)
Comment (jcosentino11) [reply src/main/java/com/aws/greengrass/mqttclient/AwsIotMqtt5Client.java]: done, had synchronize it to appease spotbugs

**Review by MikeDombo — APPROVED** [review]

**Review by robcmann — APPROVED** [review]

**Comment by MikeDombo** [comment]
Looks like OTF tests failed which may definitely be a real problem.

**Comment by jcosentino11** [comment]
> Looks like OTF tests failed which may definitely be a real problem.

OTF logs show that the client connected fine:
```
2023-11-09T00:36:16.649Z [INFO] (Thread-4) com.aws.greengrass.mqttclient.AwsIotMqtt5Client: Successfully connected to AWS IoT Core. {clientId=gg-bde16caf6d6df07d5a50-ggc-thing, sessionPresent=false}
```

only error in greengrass.log is:
```
2023-11-09T00:36:24.242Z [ERROR] (pool-2-thread-6) com.aws.greengrass.deployment.DeploymentService: Error occurred while processing deployment. {deploymentId=43056083-5a06-43a0-966a-4298f74de02b, serviceName=DeploymentService, currentState=RUNNING}
java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: software/amazon/awssdk/services/s3/model/S3Exception
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at com.aws.greengrass.deployment.DefaultDeploymentTask.call(DefaultDeploymentTask.java:147)
	at com.aws.greengrass.deployment.DefaultDeploymentTask.call(DefaultDeploymentTask.java:50)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**Comment by jcosentino11** [comment]
> > Looks like OTF tests failed which may definitely be a real problem.
> 
> OTF logs show that the client connected fine:
> 
> ```
> 2023-11-09T00:36:16.649Z [INFO] (Thread-4) com.aws.greengrass.mqttclient.AwsIotMqtt5Client: Successfully connected to AWS IoT Core. {clientId=gg-bde16caf6d6df07d5a50-ggc-thing, sessionPresent=false}
> ```
> 
> only error in greengrass.log is:
> 
> ```
> 2023-11-09T00:36:24.242Z [ERROR] (pool-2-thread-6) com.aws.greengrass.deployment.DeploymentService: Error occurred while processing deployment. {deploymentId=43056083-5a06-43a0-966a-4298f74de02b, serviceName=DeploymentService, currentState=RUNNING}
> java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: software/amazon/awssdk/services/s3/model/S3Exception
> 	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
> 	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
> 	at com.aws.greengrass.deployment.DefaultDeploymentTask.call(DefaultDeploymentTask.java:147)
> 	at com.aws.greengrass.deployment.DefaultDeploymentTask.call(DefaultDeploymentTask.java:50)
> 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> 	at java.base/java.lang.Thread.run(Thread.java:829)
> ```

Runs have mostly been passing in previous runs, starting from 6 days ago

**Comment by MikeDombo** [comment]
> > > Looks like OTF tests failed which may definitely be a real problem.
> > 
> > 
> > OTF logs show that the client connected fine:
> > ```
> > 2023-11-09T00:36:16.649Z [INFO] (Thread-4) com.aws.greengrass.mqttclient.AwsIotMqtt5Client: Successfully connected to AWS IoT Core. {clientId=gg-bde16caf6d6df07d5a50-ggc-thing, sessionPresent=false}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > only error in greengrass.log is:
> > ```
> > 2023-11-09T00:36:24.242Z [ERROR] (pool-2-thread-6) com.aws.greengrass.deployment.DeploymentService: Error occurred while processing deployment. {deploymentId=43056083-5a06-43a0-966a-4298f74de02b, serviceName=DeploymentService, currentState=RUNNING}
> > java.util.concurrent.ExecutionException: java.lang.NoClassDefFoundError: software/amazon/awssdk/services/s3/model/S3Exception
> > 	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
> > 	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
> > 	at com.aws.greengrass.deployment.DefaultDeploymentTask.call(DefaultDeploymentTask.java:147)
> > 	at com.aws.greengrass.deployment.DefaultDeploymentTask.call(DefaultDeploymentTask.java:50)
> > 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
> > 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
> > 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
> > 	at java.base/java.lang.Thread.run(Thread.java:829)
> > ```
> 
> Runs have mostly been passing in previous runs, starting from 6 days ago

I know what the problem is. We need to bump the Nucleus version in main now.

**Pull request merged**